build-clang-with-args.sh

```diff
@@ -0,0 +1,183 @@
+#
+# This requires a gcc/qemu toolchain dir made by `build-gcc-with-args.sh`
+#
+# Builds:
+# - Clang/LLVM
+#
+# Then:
+# - Appends to `buildinfo` with the clang info
+# - Adds cross-compilation configuration files for certain build systems
+# - Creates a tar file of the whole install directory
+
+set -e
+set -x
+set -o pipefail
+
+if ! [ "$#" -ge 3 ]; then
+  echo "Usage: $0 <config_name> <target> <dest_dir> <cflags...>"
+  exit 2
+fi;
+
+## Take configuration from arguments
+# This is the name for the tar file.
+# Suggested to be the gcc config with s/gcc/clang/. Will be updated if it
+# contains 'gcc'
+toolchain_name="${1}"
+# This is the expected gcc target triple (so we can set a default, and invoke gcc)
+toolchain_target="${2}"
+# This is the directory where we want the toolchain to added to
+toolchain_dest="${3}"
+# Remaining cflags for build configurations
+toolchain_cflags=("${@:4}")
+
+build_top_dir="${PWD}"
+
+# For *_VERSION variables
+# shellcheck source=sw-versions.sh
+source "${build_top_dir}/sw-versions.sh"
+
+# Fixup toolchain_name in case it includes `gcc`
+case "${toolchain_name}" in
+  *gcc*)
+    echo "Warning: Toolchain name given includes 'gcc': ${toolchain_name}";
+    toolchain_name="${toolchain_name//gcc/clang}";
+    echo "  New Toolchain name: ${toolchain_name}";
+  ;;
+esac
+
+tag_name="$(git -C "${build_top_dir}" describe --always)"
+toolchain_full_name="${toolchain_name}-${tag_name}"
+
+mkdir -p "${build_top_dir}/build"
+cd "${build_top_dir}/build"
+
+
+llvm_dir="${build_top_dir}/build/llvm-project"
+git clone https://github.com/llvm/llvm-project.git "${llvm_dir}"
+cd "${llvm_dir}"
+git checkout --force "${LLVM_VERSION}"
+
+# Clang Symlinks
+clang_links_to_create="clang++"
+clang_links_to_create+=";${toolchain_target}-clang;${toolchain_target}-clang++"
+# LLD Symlinks
+lld_links_to_create="ld.lld;ld64.lld"
+lld_links_to_create+=";${toolchain_target}-ld.lld;${toolchain_target}-ld64.lld"
+
+llvm_build_dir="${build_top_dir}/build/llvm-build"
+mkdir -p "${llvm_build_dir}"
+cd "${llvm_build_dir}"
+
+# TODO:
+# - Stage 2 Build
+# - Build compiler-rt and other runtimes
+
+cmake "${llvm_dir}/llvm" \
+  -Wno-dev \
+  -DCMAKE_C_COMPILER="/usr/bin/clang-6.0" \
+  -DCMAKE_CXX_COMPILER="/usr/bin/clang++-6.0" \
+  -DCMAKE_BUILD_TYPE=Release \
+  -DCMAKE_INSTALL_PREFIX="${toolchain_dest}" \
+  -DLLVM_TARGETS_TO_BUILD="host;RISCV" \
+  -DLLVM_ENABLE_PROJECTS="clang;lld;clang-tools-extra" \
+  -DLLVM_ENABLE_LLD=On \
+  -DLLVM_ENABLE_BACKTRACES=Off \
+  -DLLVM_DEFAULT_TARGET_TRIPLE="${toolchain_target}" \
+  -DCLANG_VENDOR="lowRISC" \
+  -DBUG_REPORT_URL="toolchains@lowrisc.org" \
+  -DLLVM_INCLUDE_EXAMPLES=Off \
+  -DLLVM_INCLUDE_DOCS=Off \
+  -DLLVM_INSTALL_TOOLCHAIN_ONLY=On \
+  -DLLVM_INSTALL_BINUTILS_SYMLINKS=Off \
+  -DCLANG_LINKS_TO_CREATE="${clang_links_to_create}" \
+  -DLLD_SYMLINKS_TO_CREATE="${lld_links_to_create}" \
+  -DLLVM_DISTRIBUTION_COMPONENTS="clang;clang-format;clang-tidy;clang-resource-headers;lld"
```

Comment:
- [ ] LLVM's binutils. (objdump, etc)

Reply:
- [ ] `scan-build` would be good to include, maybe?

Reply:
- [ ] llvm-readelf
`llvm-readelf` should be a symlink to `llvm-readobj`, as it's not longer an llvm-tool by itself. I need to check how that's currently done.